syne_tune/optimizer/schedulers/neuralbands/neuralband_supplement.py

```diff
@@ -0,0 +1,494 @@
+        if self.time_keeper is None:
+            self.time_keeper = RealTimeKeeper()
+            self.time_keeper.start_of_time()
+        # For pause/resume schedulers: Can a paused trial be promoted?
+        promote_trial_id, extra_kwargs = self._promote_trial()
+        if promote_trial_id is not None:
+            promote_trial_id = int(promote_trial_id)
+            return TrialSuggestion.resume_suggestion(
+                trial_id=promote_trial_id, config=extra_kwargs
+            )
+        # Ask searcher for config of new trial to start
+        extra_kwargs["elapsed_time"] = self._elapsed_time()
+        trial_id = str(trial_id)
+        
+        # TS selection criterion
+        initial_budget = self.net.average_b
+        while_loop_count = 0
+        l_t_score = []
+        while 1:
+            config = self.searcher.get_config(**extra_kwargs, trial_id=trial_id)
+            if config is not None:
+                config_encoding =  self.hp_ranges.to_ndarray(config)
+                predict_value = self.net.predict((config_encoding, initial_budget))
+                self.net.func.zero_grad()
+                predict_value.backward(retain_graph=True)
+                g = torch.cat([p.grad.flatten().detach() for p in self.net.func.parameters()])
+                cb = torch.sqrt(torch.sum(self.lamdba * g * g / self.U)).item()* self.nu
+                mean_value  = torch.tensor(predict_value.item())
+                predict_score = torch.normal(mean= mean_value, std=cb).item()
+                l_t_score.append((config, predict_score))
+                if self.mode == "min":
+                    if while_loop_count > self.max_while_loop:
+                        l_t_score = sorted(l_t_score, key = lambda x: x[1])
+                        config = l_t_score[0][0]
+                        break
+                else:
+                    if while_loop_count > (self.max_while_loop/20):
+                        l_t_score = sorted(l_t_score, key = lambda x: x[1], reverse=True)
+                        config = l_t_score[0][0]
+                        break
+                while_loop_count += 1   
+            else:
+                self._searcher_initialized = False
+                self._initialize_searcher_new()
+                config = self.searcher.get_config(**extra_kwargs, trial_id=trial_id)
+                break
+                                
+        if config is not None:
+            config = cast_config_values(config, self.config_space)
+            config = self._on_config_suggest(config, trial_id, **extra_kwargs)
+            config = TrialSuggestion.start_suggestion(config)
+        return config 
+    
+    
+    
+class NeuralbandUCBScheduler(NeuralbandSchedulerBase):
+    def __init__(self, config_space, lamdba = 0.01, nu = 0.01, step_size = 30, max_while_loop = 100,  **kwargs):
+        """
+        We combine Upper Confidence Bound with NeuralBand, where configuration selection criterion follows [1].
```

Comment:
We combine Upper Confidence Bound with NeuralBand, where configuration selection criterion follows [1] -> We combine Upper Confidence Bound with NeuralBand, where configuration are selected based on the upper confidence bound criterion following [1]

Reply:
Done.